core/vm/privacy/ringct.go

```diff
@@ -174,7 +175,24 @@ func (r *RingSignature) Serialize() ([]byte, error) {
 }
 
 func computeSignatureSize(numRing int, ringSize int) int {
-	return 8 + 8 + 32 + 32 + numRing*ringSize*32 + numRing*ringSize*33 + numRing*33
+	const MaxInt = int(^uint(0) >> 1)
+
+	if numRing < 0 || ringSize < 0 {
+		return -1
+	}
+
+	// Calculate each term separately and check for overflow
+	term1 := 8 + 8 + 32 + 32
+	term2 := numRing * ringSize * 32
```

Comment:
could `numRing * ringSize` overflow? could  `numRing * ringSize * 32` overflow?

Reply:
yes it could, their types are both int

Reply:
The value is actually reading from signature
```
func Deserialize(r []byte) (*RingSignature, error) {
	if len(r) < 16 {
		return nil, errors.New("Failed to deserialize ring signature")
	}
	offset := 0
	sig := new(RingSignature)
	numRing := r[offset : offset+8]
	offset += 8
	size := r[offset : offset+8]
	offset += 8

	size_uint := binary.BigEndian.Uint64(size)
	size_int := int(size_uint)
	sig.Size = size_int

Reply:
maybe we should check multiplication overflow too

Reply:
In fact, the function `computeSignatureSize` should return `uint`, not `int`. The below function signature maybe better:

```go
func computeSignatureSize(numRing uint64, ringSize uint64) uint64
```

When the compute result is overflow, we can return `0`.

Reply:
I think -1 is better since the final check is compared length, this quarantine is definitely return error
```
	if len(r) != computeSignatureSize(sig.NumRing, sig.Size) {
		return nil, fmt.Errorf("incorrect ring size, len r: %d, sig.NumRing: %d sig.Size: %d", len(r), sig.NumRing, sig.Size)
	}
```

Reply:
> maybe we should check multiplication overflow too

@wgr523  I didn't get you